### PR TITLE
[nanoflann] update to 1.5.0

### LINKED ISSUE #32055: [nanoflann] update to 1.5.0
### Library name

nanoflann

### New version number

1.5.0

### Other information that may be useful (release notes, etc...)

https://github.com/jlblancoc/nanoflann/releases/tag/v1.5.0

nanoflann 1.5.0: Released Jun 16, 2023

    API changes:
        Users of radius search should change their result placeholder type:
        std::vector<std::pair<IndexType, DistanceType>> => std::vector<nanoflann::ResultItem<IndexType, DistanceType>>. (See https://github.com/jlblancoc/nanoflann/issues/166 for the motivation of this change).
        More concise auxiliary (internal) type name:
        array_or_vector_selector -> array_or_vector.
        Remove obsolete parameter nChecks_IGNORED. Removed from SearchParams
        constructor too, so that structure has been renamed SearchParameters to
        enforce users to update the code and avoid mistakes with the order of its
        ctor parameters.
        Added method RadiusResultSet::empty()
        Template argument rename: AccesorType => IndexType (does not actually affect user code at all).
        Added concurrent tree building support, refer to KDTreeSingleIndexAdaptorParams::n_thread_build.
    Other changes:
        Macros to avoid conflicts with X11 symbols.
        Inline an auxiliary example function in case users want to use it and
        include the file in multiple translation units (Closes https://github.com/jlblancoc/nanoflann/issues/182).
        Move all benchmarking code, data, and scripts to [its own repository](https://github.com/MRPT/nanoflann-benchmark) to keep this repo as clean as possible.
        Fix "potentially uninitialized" GCC warning.
        Clarified, even more, in docs and examples, that L2 distances are squared distances.
        Removed the (with modern compilers) now useless inline keyword in class members.
        Add examples with GUI (requires [mrpt-gui](https://docs.mrpt.org/reference/latest/group_mrpt_gui_grp.html)):
            nanoflann_gui_example_R3: Radius search on R³ Euclidean space.
            nanoflann_gui_example_bearings: NN search on non-Euclidean spaces.
    BUGFIXES:
        Avoid segfault if saving an empty index (Closes https://github.com/jlblancoc/nanoflann/issues/205).

### PR DESCRIPTION
**Pull request by MonicaLiu0311**
Fixes #32055. 

* Update `nanoflann` to 1.5.0.
* The following error occurred when testing usage, so add dependency `Threads` for `nanoflannConfig.cmake`.
```
1> [CMake] CMake Error at E:/vcpkg01/installed/x64-windows/share/nanoflann/nanoflannTargets.cmake:60 (set_target_properties):
1> [CMake]   The link interface of target "nanoflann::nanoflann" contains:
1> [CMake] 
1> [CMake]     Threads::Threads
1> [CMake] 
1> [CMake]   but the target was not found.  Possible reasons include:
1> [CMake] 
1> [CMake]     * There is a typo in the target name.
1> [CMake]     * A find_package call is missing for an IMPORTED target.
1> [CMake]     * An ALIAS target is missing.
1> [CMake] 
1> [CMake] Call Stack (most recent call first):
1> [CMake]   E:/vcpkg01/installed/x64-windows/share/nanoflann/nanoflannConfig.cmake:27 (include)
1> [CMake]   E:/vcpkg01/scripts/buildsystems/vcpkg.cmake:853 (_find_package)
1> [CMake]   CMakeFindUsage/CMakeLists.txt:15 (find_package)
1> [CMake] -- Generating done (0.0s)
```
* [The API changes](https://github.com/jlblancoc/nanoflann/blob/v1.5.0/CHANGELOG.md?plain=1#L2-L13) of `nanoflann` caused two errors when installing the downstream `pdal`, which were fixed in `fix-dependency.patch`. Reported upstream: https://github.com/PDAL/PDAL/issues/4081
```
E:\vcpkg01\buildtrees\pdal\src\2.5.3-f9d28ae14f\pdal\private/KDImpl.hpp(100): error C2039: 'SearchParams': is not a member of 'nanoflann'
E:\vcpkg01\buildtrees\pdal\src\2.5.3-f9d28ae14f\pdal\private/KDImpl.hpp(100): error C3861: 'SearchParams': identifier not found
```
```
E:\vcpkg01\buildtrees\pdal\src\2.5.3-f9d28ae14f\pdal\private/KDImpl.hpp(128): error C2664: 'unsigned __int64 nanoflann::KDTreeSingleIndexAdaptor<nanoflann::L2_Simple_Adaptor<double,pdal::KD2Impl,double,uint32_t>,pdal::KD2Impl,-1,size_t>::radiusSearch(const double *,const double &,std::vector<nanoflann::ResultItem<IndexType,double>,std::allocator<nanoflann::ResultItem<IndexType,double>>> &,const nanoflann::SearchParameters &) const': cannot convert argument 3 from 'std::vector<std::pair<size_t,double>,std::allocator<std::pair<size_t,double>>>' to 'std::vector<nanoflann::ResultItem<IndexType,double>,std::allocator<nanoflann::ResultItem<IndexType,double>>> &'
        with
        [
            IndexType=size_t
        ]
```
* No feature needs to be tested, the usage test passed (header files found): 
```
nanoflann provides CMake targets:

    # this is heuristically generated, and may not be correct
    find_package(nanoflann CONFIG REQUIRED)
    target_link_libraries(main PRIVATE nanoflann::nanoflann)
```

- [X] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [X] SHA512s are updated for each updated download
- [ ] ~~The "supports" clause reflects platforms that may be fixed by this new version~~
- [ ] ~~Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.~~
- [ ] ~~Any patches that are no longer applied are deleted from the port's directory.~~
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

File: ports/nanoflann/portfile.cmake [hunk 1]
```diff
@@ -11,6 +11,8 @@ vcpkg_cmake_configure(
     OPTIONS
         -DNANOFLANN_BUILD_EXAMPLES=OFF
         -DNANOFLANN_BUILD_TESTS=OFF
+    MAYBE_UNUSED_VARIABLES
+        PKG_CONFIG_EXECUTABLE
```
Comment (dg0yt) [inline ports/nanoflann/portfile.cmake:15]: Wrong portfile? I can't see how `PKG_CONFIG_EXECUTABLE` could ever be set here.

Comment (MonicaLiu0311) [reply ports/nanoflann/portfile.cmake]: Done, thanks.

**Review by LilyWangLL — APPROVED** [review]

**Comment by hobu** [comment]
I don't understand the need for this patch to PDAL. While we haven't updated the vendored nanoflann inside of PDAL for two+ years, it shouldn't be necessary to update our KDtree implementation to use a newer system-available nanoflann. Something is causing the system-available nanoflann to be scooped up as an include before the vendored one. Some local testing shows this doesn't happen for me on OSX with a condaforge install of a public nanoflann.

**Comment by dg0yt** [comment]
@hobu It is very simple: The vcpkg port replaces the vendored copy with vcpkg's nanoflann port.

**Comment by hobu** [comment]
> The vcpkg port replaces the vendored copy with vcpkg's nanoflann port.

Ok, [Monica's ticket](
https://github.com/PDAL/PDAL/issues/4081) wasn't explicit the packager was *swapping* out nanoflann instead of implying that there was some kind of confusion with a system-installed nanoflann and what PDAL has vendored. This caused me to spend time to investigate the issue. I would not have bothered if I had known you were essentially trying to update the vendored version.

As for updating the vendored nanoflann, I tried this about 10 months ago and found the new version was 30% slower for PDAL's usage. Package systems like vcpkg having policy to go run the last version of whatever vendored software a package might provide is not always the best policy. This is one of those cases.

**Comment by hobu** [comment]
>  As far as I can see PDAL doesn't move the vendored copy's symbols into a private namespace. I

We do that for nlohmann, which PDAL also vendors. I would expect it wouldn't be difficult to do that for nanoflann, and I have ticketed it at https://github.com/PDAL/PDAL/issues/4085

> I hope you reported the performance problem to nanoflann.

No I didn't. nanoflann's recent development tact has kind of de-emphasized the *nano* part 😄 Our vendored nanoflann usage had been working fine for years, so we were happy to leave it alone.

**Comment by hobu** [comment]
> As far as I can see PDAL doesn't move the vendored copy's symbols into a private namespace.

Can you show me how this is true? My investigation showed that we bury the nanoflann usage in a private class that isn't exported https://github.com/PDAL/PDAL/blob/master/pdal/private/KDImpl.hpp

**Comment by dg0yt** [comment]
I doubt I can show this. What I tried to say before is that there may be no actual problem here in the particular case of PDAL. It has a combination of dynamic linkage, private classes and hidden visibility by default.
In general, there are good reasons to be concerned about vendored copies which haven't been updated in three years, and about packages which come with vendored copies of multiple packages.

**Pull request merged**